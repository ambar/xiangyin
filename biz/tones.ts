/**
 * https://zh.wikipedia.org/wiki/四聲
 */

export const toneCodeByName = {
  T0: '轻声',
  T1: '平声',
  T1a: '阴平',
  T1b: '阳平',
  T2: '上声',
  T2a: '阴上',
  T2b: '阳上',
  T3: '去声',
  T3a: '阴去',
  T3b: '阳去',
  T4: '入声',
  T4a: '阴入',
  T4b: '阳入',
}

export const toneNameByCode = {
  轻声: 'T0',
  平: 'T2',
  平声: 'T2',
  阴平: 'T1a',
  阳平: 'T1b',
  上: 'T2',
  上声: 'T2',
  阴上: 'T2a',
  阳上: 'T2b',
  去: 'T3',
  去声: 'T3',
  阴去: 'T3a',
  阳去: 'T3b',
  入: 'T4',
  入声: 'T4',
  阴入: 'T4a',
  阳入: 'T4b',
}

export const csToneNoByCode = {
  T0: '0',
  T1a: '1',
  T1b: '2',
  T2: '3',
  T3a: '4',
  T3b: '5',
  T4: '6',
}

export const csToneNoByTone = {
  '1': 1,
  '2': 2,
  '3': 3,
  '5': 4,
  '6': 5,
  '7': 6,
}
