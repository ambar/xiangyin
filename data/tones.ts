/**
 * 声调格式转换
 *
 * @see https://zh.wikipedia.org/wiki/四聲
 */

export type ToneType = 'ToneName' | 'CSToneNo' | 'OctetToneNo'

export type ToneCode = keyof typeof toneCode2toneName
export const toneCode2toneName = {
  T0: '轻声',
  T1: '平声',
  T1a: '阴平',
  T1b: '阳平',
  T2: '上声',
  T2a: '阴上',
  T2b: '阳上',
  T3: '去声',
  T3a: '阴去',
  T3b: '阳去',
  T4: '入声',
  T4a: '阴入',
  T4b: '阳入',
}

export type ToneName = keyof typeof toneName2ToneCode
export const toneName2ToneCode = {
  轻声: 'T0',
  // 平: 'T2',
  平声: 'T2',
  阴平: 'T1a',
  阳平: 'T1b',
  // 上: 'T2',
  上声: 'T2',
  阴上: 'T2a',
  阳上: 'T2b',
  // 去: 'T3',
  去声: 'T3',
  阴去: 'T3a',
  阳去: 'T3b',
  // 入: 'T4',
  入声: 'T4',
  阴入: 'T4a',
  阳入: 'T4b',
}

/** 长沙话序号声调 */
export type CSToneNo = 1 | 2 | 3 | 4 | 5 | 6

/** 长沙话序号声调 */
export const toneCode2csToneNo: Partial<Record<ToneCode, CSToneNo | ''>> = {
  T0: '',
  T1a: 1,
  T1b: 2,
  T2: 3,
  T3a: 4,
  T3b: 5,
  T4: 6,
  T4a: 6,
}

/** 八位数字声调 */
export type OctetToneNo = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8
/** 长沙话在用的八位数字声调  */
export type CSOctetToneNo = Exclude<OctetToneNo, 4 | 8>

/** 八位数字声调 */
export const octetToneNo2csToneNo: Record<CSOctetToneNo, number> = {
  1: 1,
  2: 2,
  3: 3,
  5: 4,
  6: 5,
  7: 6,
}

export const csOctetToneNo2toneName: Record<CSOctetToneNo, ToneName> = {
  1: '阴平',
  2: '阳平',
  3: '上声',
  5: '阴去',
  6: '阳去',
  7: '入声',
}

/** 四声转长沙话序号 */
export const csToneName2csToneNo: Partial<Record<ToneName, CSToneNo | ''>> = {
  轻声: '',
  阴平: 1,
  阳平: 2,
  上声: 3,
  阴去: 4,
  阳去: 5,
  入声: 6,
}

/** 调值转长沙话序号 */
export const toneValue2csToneNo: Record<number, CSToneNo> = {
  33: 1,
  13: 2,
  41: 3,
  55: 4,
  11: 5,
  24: 6,
}
