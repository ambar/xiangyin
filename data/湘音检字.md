# 湘音检字

## 修正

- 去除了重复的行（3K+）
- 「攡」应当在字甲

## 问题

文件默认包含了一些简体/异体对应，但不全面，一些例外：

```text
弄 衖
纤 纖
减 減
们 們
础 礎
迁 遷
团 團
丛 叢
导 導
头 頭
触 觸
沣 灃
尝 嘗
```

## 参考

- [《湘音检字》与长沙方言 - 鲍厚星](https://www.zhihu.com/xen/market/remix/paid_magazine/1395144375069016064)
- [《湘音检字》及其反映的三套语音系统 - 田范芬](https://www.zhihu.com/xen/market/remix/paid_magazine/1395738931791511552)
- [两本长沙方言韵书:《训诂谐音》和《湘音检字》 - 田范芬](https://www.zhihu.com/xen/market/remix/paid_magazine/1394207942187569152)
