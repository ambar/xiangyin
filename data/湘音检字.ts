import {createJyinEntry} from './jyin'
import json from './raw/湘音检字.json'
import {changeTone, CSToneNo} from './tones'
import {JyinEntry, SchemaEntries} from './types'
import {Final, Finals, Initial, Initials} from './湘音检字.meta'
export * from './湘音检字.meta'

const schema = [
  ['湘拼', String],
  ['音標', String],
  ['調號', Number],
  ['字甲', String],
  ['字乙', String],
  ['釋義', String],
] as const

type DataItem = SchemaEntries<typeof schema> & {
  声: Initial
  韵: Final
  规范: JyinEntry
}

// 异体、多音字分组 16519 -> 13543
export const charGroup = new Map<string, DataItem[]>()

// 声母较长的排前面
const sortedInitials = Object.entries(Initials).sort(
  (a, b) => b[1].length - a[1].length
)

/** 分解原文 IPA */
export const ipa2senyn = (ipa: string) => {
  let i = ''
  let f = ''
  // 先从零声母/韵母查起，因为 m̩（韵母）以 m（己有声母）开头
  const matchF = ipa in Finals
  if (matchF) {
    i = ''
    f = ipa
  } else {
    const matchSort = sortedInitials.find((x) => ipa.startsWith(x[1]))
    if (matchSort) {
      i = matchSort[0]
      f = i === '' ? ipa : ipa.replace(i, '')
    }
  }
  return [i, f] as [Initial, Final]
}

export const items = json.map((x) => {
  const item = Object.fromEntries(
    schema.map(([key, val], i) => [key, val(x[i])])
  ) as unknown as DataItem
  const setIfNotSet = (char: string) => {
    if (char) {
      if (charGroup.has(char)) charGroup.get(char)!.push(item)
      else charGroup.set(char, [item])
    }
  }
  const [声, 韵] = ipa2senyn(item.音標)
  item.声 = 声
  item.韵 = 韵
  item.规范 = createJyinEntry(
    item.声,
    item.韵,
    changeTone(item.調號, 'OctetToneNo', 'CSToneNo') as CSToneNo,
    item.釋義
  )
  setIfNotSet(item.字甲)
  setIfNotSet(item.字乙)
  return item
})

export const query = (char: string): JyinEntry[] => {
  const items = charGroup.get(char) || []
  return items.map((item) => item.规范)
}
