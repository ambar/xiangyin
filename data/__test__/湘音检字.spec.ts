import {ToneType, ToneTypes} from '../tones'
import {query} from '../湘音检字'

test('query', () => {
  expect(query('中')).toMatchSnapshot()
  const zhy = query('鑄')
  expect(query('鋳')).toEqual(zhy)
  expect(query('铸')).toEqual(zhy)
})

test.each(ToneTypes.concat('XYZ' as ToneType))('toneType: %s', (toneType) => {
  expect(query('入', toneType)).toMatchSnapshot()
})
