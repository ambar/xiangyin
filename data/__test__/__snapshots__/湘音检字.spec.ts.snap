// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ipa2senyn: kai 1`] = `
Array [
  Array [
    "k",
    "ai",
  ],
  Array [
    "g",
    "ai",
  ],
]
`;

exports[`ipa2senyn: m̩ 1`] = `
Array [
  Array [
    "",
    "m̩",
  ],
  Array [
    "",
    "m",
  ],
]
`;

exports[`ipa2senyn: pã 1`] = `
Array [
  Array [
    "p",
    "ã",
  ],
  Array [
    "b",
    "an",
  ],
]
`;

exports[`ipa2senyn: ɿ 1`] = `
Array [
  Array [
    "",
    "ɿ",
  ],
  Array [
    "r",
    "",
  ],
]
`;

exports[`query 1`] = `
Array [
  Object {
    "声": "tʃ",
    "调": 1,
    "释": "",
    "音": "tʃoŋ",
    "韵": "oŋ",
  },
  Object {
    "声": "tʃ",
    "调": 4,
    "释": "",
    "音": "tʃoŋ",
    "韵": "oŋ",
  },
]
`;
